### PR TITLE
[blog] add ghost blogging infrastructure to blog.hail.is

### PR DESCRIPTION
**Pull request by catoverdrive**
this deploys a standard ghost installation to the default namespace and routes to it from blog.hail.is; can be customized once deployed.

This won't pass tests until #7413 is in.

### CODE REVIEW

File: blog/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  template:
+    metadata:
+      labels:
+        app: blog
+    spec:
+{% if deploy %}
+      priorityClassName: production
+{% endif %}
+      containers:
+      - name: blog
```
Comment (cseed) [inline blog/deployment.yaml]: You should include liveness and readiness probes.  You'll need to change the port and URL for ghost:

```
          livenessProbe:
            httpGet:
              path: /healthcheck
              port: 5000
            initialDelaySeconds: 5
            periodSeconds: 5
          readinessProbe:
            httpGet:
              path: /healthcheck
              port: 5000
            initialDelaySeconds: 5
            periodSeconds: 5
```

These are important so if the service gets stuck, k8s will restart, and when upgrading, k8s won't route traffic to the service until it can respond.  (Although it doesn't really matter for a single-instance stateful set, it is still best practice.)

File: router/router.nginx.conf.in [hunk 2]
```diff
@@ -232,3 +232,19 @@ server {
...
+    location / {
+        proxy_pass http://blog/;
+
+        proxy_set_header Host $updated_host;
+        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+    }
+
+    listen 80;
+    listen [::]:80;
+	client_max_body_size 8M;
```
Comment (cseed) [inline router/router.nginx.conf.in]: Why did you add this?  It seems quite large.

File: blog/deployment.yaml (hunk 1 above)
Comment (cseed) [reply blog/deployment.yaml]: You should also include resources.  `requests` is the minimum resources that should be reserved for it, for example:

```
          resources:
            requests:
              memory: "1G"
              cpu: "300m"
```

`limits`, optional, is the maximum amount of resources the service can take advantage of.  We probably won't get a ton of traffic, so a fraction of a CPU like 300m is probably fine.  We'll want to revisit if we e.g. post something on hacker news.  I normally set memory by running the service for a while and watching how much memory it uses.  You can get the memory usage with `kubectl top pods`.

File: router/router.nginx.conf.in (hunk 2 above)
Comment (akotlar) [reply router/router.nginx.conf.in]: This would be useful for image/video uploads; we should set this to whatever the largest file size we want for media.

Comment (danking) [reply router/router.nginx.conf.in]: I don't think this will make it past the gateway though

Comment (catoverdrive) [reply router/router.nginx.conf.in]: I ran into this when trying to upload a 1.2MB theme. Adding this line worked for the dev deployed version; would gateway do something else for blog.hail.is?

And yeah, I set this semi-arbitrarily with an image/video upload limit in mind. Discord does ~7MB per post and I thought that was pretty reasonable.

File: build.yaml [hunk 3]
```diff
@@ -1070,3 +1070,15 @@ steps:
...
     - make_docs
     - build_hail
+ - kind: deploy
+   name: deploy_blog
+   namespace:
+     valueFrom: default_ns.name
+   config: blog/deployment.yaml
+   scopes:
+    - deploy
+    - dev
```
[the remark this answers is not in the document]
Comment (catoverdrive) [reply build.yaml]: The service is started in deploy_router, though. Is that still useful to have here?

[the remark this answers is not in the document]
Comment (catoverdrive) [reply build.yaml]: ah, the healthcheck makes sure that there's something that's actually pods backing the service?

File: build.yaml [hunk 4]
```diff
@@ -1070,3 +1070,16 @@ steps:
     - base_image
     - make_docs
     - build_hail
+ - kind: deploy
+   name: deploy_blog
+   namespace:
+     valueFrom: default_ns.name
```
Comment (akotlar) [inline build.yaml]: this should be {{ default_ns.name }}

Comment (akotlar) [reply build.yaml]: Ah nope, nvm

**Comment by akotlar** [comment]
I think your error is happening because you're requesting a resource that doesn't exist

`kubectl -n pr-7381-default-vo6ftnzp8ta2 rollout status --timeout=1h deployment blog`

You need "statefulset" here:

Example with a similar kind of rollout:

```sh
$ k -n monitoring rollout status --timeout=1h deployment prometheus
Error from server (NotFound): deployments.extensions "prometheus" not found

$ k -n monitoring rollout status --timeout=1h statefulset prometheus
partitioned roll out complete: 1 new pods have been updated...
```

**Comment by akotlar** [comment]
I think this is happening from this line in ci/ci.py. When ci finds a `wait` step of `kind: Service`, it's rollout command checks a deployment, which we of course don't have. We may need to allow statefulsets here, either through a config property, or by checking whether the deploy `config` for the build step has a `StatefulSet`, or `Deployment`.

```python
if self.wait:
            for w in self.wait:
                if w['kind'] == 'Pod':
                    script += f'''\
kubectl -n {self.namespace} delete --ignore-not-found pod {w['name']}
'''
        script += f'''
echo {shq(rendered_config)} | kubectl -n {self.namespace} apply -f -
'''

        if self.wait:
            for w in self.wait:
               # ... redacted ...
                elif w['kind'] == 'Service':
                    assert w['for'] == 'alive', w['for']
                    port = w.get('port', 80)
                    timeout = w.get('timeout', 60)
                    script += f'''
set +e
kubectl -n {self.namespace} rollout status --timeout=1h deployment {name} && \
  kubectl -n {self.namespace} wait --timeout=1h --for=condition=available deployment {name} && \
  python3 wait-for.py {timeout} {self.namespace} Service -p {port} {name}
EC=$?
kubectl -n {self.namespace} logs --tail=999999 -l app={name} | {pretty_print_log}
set -e
(exit $EC)
'''
```

**Comment by catoverdrive** [comment]
I'm also kind of unsure how to implement the liveness/readiness probes. Ghost doesn't really have a healthcheck endpoint, and doing a get request on `/` for the probe seems... suboptimal?

**Comment by danking** [comment]
Why does that seem sub-optimal? That seems right for readiness. If it can serve the root, it's probably ready to go.

For liveness, ideally the page doesn't introduce too much load because it gets hit somewhat frequently.

None of this really matters until we have traffic anyway. If we anticipate a surge, we can make sure these aren't overloading the system before the surge.

**Comment by catoverdrive** [comment]
ok, I'll do that.

File: ci/ci/build.py [hunk 5]
```diff
@@ -680,11 +680,12 @@ def build(self, batch, code, scope):
...
                 elif w['kind'] == 'Service':
                     assert w['for'] == 'alive', w['for']
+					resource_type = w.get("resource_type", "deployment")
                     port = w.get('port', 80)
                     timeout = w.get('timeout', 60)
                     script += f'''
 set +e
-kubectl -n {self.namespace} rollout status --timeout=1h deployment {name} && \
+kubectl -n {self.namespace} rollout status --timeout=1h {resource_type} {name} && \
   kubectl -n {self.namespace} wait --timeout=1h --for=condition=available deployment {name} && \
```
Comment (akotlar) [inline ci/ci/build.py]: needed here too

Comment (akotlar) [reply ci/ci/build.py]: Unfortunately, I don't think this will uniformly work. 

https://github.com/kubernetes/kubernetes/issues/79606

The rollout command works fine, but when I wait --for=condition=available on grafana, it will never find the condition met.

Comment (akotlar) [reply ci/ci/build.py]: This may be fixed in 1.12 (https://github.com/kubernetes/kubernetes/pull/62943), and we're in 1.14. Helm people still seem to be having issues: https://github.com/helm/helm/issues/4539

File: router/router.nginx.conf.in (hunk 2 above)
Comment (danking) [reply router/router.nginx.conf.in]: Gateway has this for internal URLs, which is honestly a little unsettlingly large?
```
server {
    server_name internal.hail.is;
    client_max_body_size 10000m;
```
You'll need a similar thing under:
```
server {
    server_name hail.is;
```
but maybe do 8M instead of 10000m?

Comment (akotlar) [reply router/router.nginx.conf.in]: We wouldn't run into this for a while, but users using the eventual service may serialize IR's that are quite large right? We could enable larger body sizes for just the IR routes (to say /api/ with method POST)

Comment (akotlar) [reply router/router.nginx.conf.in]: I think the original reason this was set was to allow for very large batches to be submitted in a single request, which we moved away from, so I agree that 10GB is much more than we need, at least until we have a specific use case (uploading could be one, if we ever allow users to submit files to the cluster, rather than S3/GCS).

Comment (danking) [reply router/router.nginx.conf.in]: I assume HTTP body limits have a good reason to exist. The client of the Hail service is not a web browser, so it can break it's messages into arbitrarily small chunks.

**Comment by akotlar** [comment]
For the CI issue, without a deployment this will still break. The "rollout" portion of the code that you had would work. The "wait" portion would not, because it appears to be broken on statefulsets. I think this may work in that case:

`kubectl -n blog wait --timeout=1h --for=condition=ready pods --selector=app=blog`

Tested with `pods --all`, `pods --selector=app=prometheus`. A difference that is worth being aware of: timeout will apply to each resource matched by the selector.

So you'd need to check if the resource specified was a stateful set, and substitute a selector that they would need to pass in the config.

File: blog/deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        - containerPort: 2368
+        livenessProbe:
+          httpGet:
+{% if deploy %}
+            path: /
+{% else %}
+            path: /{{ default_ns.name }}/blog/
+{% endif %}
+            port: 2368
+          initialDelaySeconds: 5
```
Comment (akotlar) [inline blog/deployment.yaml]: timeoutSeconds defaults to 1. that is probably sufficient, but I wanted to mention that it may be useful to set this number to something higher, or just keep an eye on response times.

File: blog/deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+      labels:
+        app: blog
+    spec:
+{% if deploy %}
+      priorityClassName: production
+{% endif %}
+      containers:
+      - name: blog
+        image: ghost:3.0-alpine
+        imagePullPolicy: Always
```
Comment (akotlar) [inline blog/deployment.yaml]: Is there a reason for this to be Always over IfNotPresent?

Comment (catoverdrive) [reply blog/deployment.yaml]: I don't believe so. I can change it; should I just remove that line and have it use the default?

File: blog/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+{% endif %}
+      containers:
+      - name: blog
+        image: ghost:3.0-alpine
+        env:
+          - name: NODE_ENV
+            value: production
+          - name: url
+{% if deploy %}
+            value: http://blog.hail.is
```
Comment (akotlar) [inline blog/deployment.yaml]: I think these should be https

File: blog/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+            value: http://internal.hail.is/{{ default_ns.name }}/blog
+{% endif %}
+        ports:
+        - containerPort: 2368
+        livenessProbe:
+          httpGet:
+{% if deploy %}
+            path: /
+{% else %}
+            path: /{{ default_ns.name }}/blog/
```
Comment (akotlar) [inline blog/deployment.yaml]: If I set the ghost url to https://domain.com/path/to/subfolder/ is root / or https://domain.com/path/to/subfolder/?

File: blog/deployment.yaml (hunk 8 above)
Comment (catoverdrive) [reply blog/deployment.yaml]: Since we're handling the ssl/certificate stuff at the gateway level and using entirely http behind the gateway, I think these have to be http connections to work with our nginx setup. The only annoyance is that links on the website will default to e.g. `http://blog.hail.is/post/foo` which will redirect to https once they hit our gateway, but I don't think we can/want to have the ghost server think it should be using https.

Comment (catoverdrive) [reply blog/deployment.yaml]: (setting the url to be https disables unsecured http connections, which we want to be using within our cluster.)

File: blog/deployment.yaml (hunk 9 above)
Comment (catoverdrive) [reply blog/deployment.yaml]: I believe Ghost's server listens on `/path/to/subfolder/`, if I understand your question/terminology correctly? Basically, if I set the ghost URL to `https://domain.com/path/to/subfolder/`, and I do (from the same machine that the ghost's server is on):
```
curl -v localhost:2368
curl -v localhost:2368/path/to/subfolder/
```
I get a 404 error for the first call, and the second call will return the homepage.

File: blog/deployment.yaml (hunk 8 above)
Comment (akotlar) [reply blog/deployment.yaml]: I think that https should work fine. It will be stripped in gateway (~line 42) inside of proxy pass. Gateway also has a server block that listens on port 80 and redirects all requests to port 443/https with a 301 redirect

```
server {
    listen 80 default_server;
    listen [::]:80 default_server;
    server_name _;

    location /.well-known/acme-challenge {
        proxy_pass http://letsencrypt;

        proxy_set_header Host $http_host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Host $http_host;
        proxy_set_header X-Forwarded-Proto https;
    }

    location / {
        return 301 https://$host$request_uri;
    }
}
```

Comment (akotlar) [reply blog/deployment.yaml]: (Since `https://blog.hail.is` would be the public url / what people click on)

File: blog/deployment.yaml (hunk 9 above)
Comment (akotlar) [reply blog/deployment.yaml]: Gotcha, thanks!

File: blog/deployment.yaml (hunk 8 above)
Comment (catoverdrive) [reply blog/deployment.yaml]: The problem with telling ghost that it has an https address isn't that https doesn't work externally (it does) but what ghost does with that information. If I specify the URL as an https URL, the Ghost server won't allow normal http connections at all, which I don't believe is the desired behavior. 

I think I want all incoming external traffic to be redirected to https when it hits our gateway, and I want all of our internal traffic behind the gateway (within our cluster, behind the ssl layer) to be plain http. if I tell Ghost that its address is https, I can (at best) set up a redirect for all incoming http traffic that hits the server to go to https, but I can't tell it to also permit requests via http. I don't think it's necessary to set up the https redirect at the Ghost server level, and it also messes with the liveness/readiness probe setup---if I try to send a http get request to a ghost server with https enabled, I'll get a 301 response instead of a 200, which fails the health check.

Comment (catoverdrive) [reply blog/deployment.yaml]: The unfortunate side effect is that Ghost now thinks it should use http://blog.hail.is/ when creating links, but that's redirected when the traffic from those links hits the gateway, so I think that's probably fine.

Comment (catoverdrive) [reply blog/deployment.yaml]: (it's kind of annoying that specifying the URL informs so much of ghost's default behavior when all I really want is an option that says "here's the prefix you should use when linking to yourself, please trust me to handle the rest", but afaict we don't really get more granularity in configuration)

Comment (danking) [reply blog/deployment.yaml]: I think Alex is referencing the solution here: https://forum.ghost.org/t/ssl-configured-ghost-generates-a-few-non-https-links-in-the-body/5888/6, if we specify `x-forwarded-proto` then Ghost will allow non-SSL connections.

Comment (catoverdrive) [reply blog/deployment.yaml]: I don't see how this helps me with the healthcheck issue, which afaict doesn't go through the nginx server? Or were you suggesting something else?

Comment (catoverdrive) [reply blog/deployment.yaml]: oh wait hmmm

Comment (catoverdrive) [reply blog/deployment.yaml]: I can set the headers in the probes as well, apparently, which should hopefully fix the issue there.

So if I'm understanding this correctly, I want to have the router's nginx config for the blog.* also set that header, right? So that'll ensure that all traffic that hits a Ghost server that I've deployed in any namespace will have X-Forwarded-Proto=https in the header which will prevent the redirect?

Comment (catoverdrive) [reply blog/deployment.yaml]: @danking

Comment (catoverdrive) [reply blog/deployment.yaml]: ok, cool. I think I set it up correctly and everything *appears* to be working with the https url. I don't think there's any difference in configuration between the internally deployed configuration and the configuration that blog.hail.is would use that I need to account for, right?

File: router/router.nginx.conf.in [hunk 10]
```diff
@@ -232,3 +232,20 @@ server {
...
+
+server {
+    server_name blog.*;
+
+    location / {
+        proxy_pass http://blog/;
+
+        proxy_set_header Host $updated_host;
+        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+        proxy_set_header X-Forwarded-Proto https;
```
Comment (akotlar) [inline router/router.nginx.conf.in]: I think this would be slightly cleaner as `$updated_scheme` instead of `https`, so that we can avoid hardcoding protocols (not a big deal here since we will likely never want anything besides https). $updated_scheme is slightly confusingly named, but represents the passed-through $scheme (protocol) from gateway ($scheme is the current request protocol, which will be https for both http and https requests, since http requests are always redirected to https).


Ghost also recommends you set the X-Forwarded-Host header: you can do so by writing `proxy_set_header X-Forwarded-Host $updated_host`. I don't think you will have a problem until we try to run ghost behind internal.hail.is (for dev purposes), because gateway sets the proxied request's HOST to $http_host when running on hail.is (but for internal.hail.is sets the Host to $service.internal where service comes from internal.hail.is/namespace/{service}). I'm assuming ghost defaults to the Host when X-Forwarded-Host isn't set.

https://ghost.org/faq/change-configured-site-url/#setup-ssl

File: build.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (akotlar) [reply build.yaml]: Yeah exactly. The readiness check runs until readiness is found (and not again), the liveness one runs in intervals.

File: blog/deployment.yaml (hunk 7 above)
Comment (akotlar) [reply blog/deployment.yaml]: yep, removing it is the right action, default is IfNotPresent

File: blog/deployment.yaml (hunk 8 above)
Comment (akotlar) [reply blog/deployment.yaml]: Yep, I think this is correct. Only relevant differences are that Host will be 1) $service.internal, and that the path will have 2) /namespace/service. You've accounted for both!

**Comment by akotlar** [comment]
Regarding the X-Forwarded-Host.

This is why the current setup works: 
(from express request lib)
https://github.com/iconoeugen/express/blob/master/lib/request.js#L432

If X-Forwarded-Host is not set, the Host header is checked. So while not strictly necessary, until Ghost stops recommending X-Forwarded-Host, I would keep it, except with a slight change.

Also, I think our use of Host is slightly wrong. From what I can tell, Host really should be the virtual/proxied host, and X-Forwarded-Host the host of the original recipient.

By having $updated_host take either of these values (Host when X-Forwarded-Host isn't set, else X-Forwarded-Host) we slightly confuse the issue. This can also lead to bugs, when X-Forwarded-Host contains multiple values (which is in spec): https://github.com/kubernetes/ingress-nginx/issues/2463

So I would use the following config for Host:

```
proxy_set_header Host $http_host
proxy_set_header X-Forwarded-Host $http_x_forwarded_host
```

However, I'm also ok with you keeping it as you have it, because that's what is done elsewhere, which will increase the likelihood that we find that solution insufficient if it truly isn't.

File: router/router.nginx.conf.in [hunk 11]
```diff
@@ -232,3 +232,20 @@ server {
...
+
+server {
+    server_name blog.*;
+
+    location / {
+        proxy_pass http://blog/;
+
+        proxy_set_header Host $updated_host;
+        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+        proxy_set_header X-Forwarded-Proto $updated_scheme;
```
Comment (danking) [inline router/router.nginx.conf.in]: Do either of you know where `update_host` and `updated_scheme` come from? I've never seen that before. I've always used `$host` and `$scheme`. I can't find any references to these variables on the internet.

Comment (akotlar) [reply router/router.nginx.conf.in]: Yeah, it's totally something we created. I think it may have been inspired by Kubernetes use of it. 

It's at the top of router.nginx.conf.in. The name is probably not quite right: it refers to the original host (as set by the upstream server in X-Forwarded-Host, X-Forwarded-Proto)

Comment (akotlar) [reply router/router.nginx.conf.in]: I think the reason cotton used this is that by default aiohttp doesn't support X-Forwarded-Host (there is a module, aiohttp-remote that addresses this), and so he stuffs that into Host, falling back to $http_host if $http_x_forwarded_host is empty.

Comment (akotlar) [reply router/router.nginx.conf.in]: This create a polymorphism that adds a bit of complexity, and potential source of bugs (if X-Forwarded-Host has multiple values, which is allowed by the spec), while also saving some complexity (not needing to look at a 2nd header field).

File: blog/deployment.yaml (hunk 8 above)
Comment (catoverdrive) [reply blog/deployment.yaml]: What does it mean for the host to be $service.internal here? I don't quite understand where that's used.

**Comment by catoverdrive** [comment]
Thanks for the explanation! I'm happy to make the change, I was just trying to understand the difference between Host and X-Forwarded-Host a little better before first.

So if I understand correctly, for the different headers:
- X-Forwarded-Proto gets passed to the router through the base https server in gateway, which sets X-Forwarded-Proto to `$scheme`, which is always going to be https since that's always going to be the protocol you're using for that server? And so when we use `$updated_scheme` for the blog server in the router's config, it's going to look at `$http_x_forwarded_proto` which will always have been set to `https` from the gateway? I'm having trouble seeing when `$http_x_forwarded_proto` would ever be absent, although if it is, isn't $scheme always `http` since all traffic from gateway to router is via http? Or am I misunderstanding how this works?
- I'm having trouble understanding the difference between `Host` and `X-Forwarded-Host`, still. As I understand it, `Host` is the name of the server that the current request is trying to reach, and `X-Forwarded-Host` is the name of the server that the original request was trying to reach? Which is why `Host` is set to `$service.internal` and `X-Forwarded-Host` is `$http_host` in the internal.hail.is server? I don't quite follow your comment about our use of `Host` being wrong, in this case; I *think* I understand what you're saying? but I'm not sure why all of our stuff is setting `Host` to `$updated_host` if that's the case, and I don't understand what's happening enough to know what happens if I change it to `proxy_set_header Host $http_host;` instead.

File: blog/deployment.yaml (hunk 8 above)
Comment (akotlar) [reply blog/deployment.yaml]: When gateway receives a request to internal.hail.is it matches on a regex pattern (something like /([^\/]+)/([^\/]+) ), that corresponds to /namespace/service.

It thens sets the proxied request's Host field to $service.internal (in your case it would be blog.internal).

**Comment by akotlar** [comment]
> Thanks for the explanation! I'm happy to make the change, I was just trying to understand the difference between Host and X-Forwarded-Host a little better before first.
> 
> So if I understand correctly, for the different headers:
> 
> * X-Forwarded-Proto gets passed to the router through the base https server in gateway, which sets X-Forwarded-Proto to `$scheme`, which is always going to be https since that's always going to be the protocol you're using for that server? And so when we use `$updated_scheme` for the blog server in the router's config, it's going to look at `$http_x_forwarded_proto` which will always have been set to `https` from the gateway? I

Yep. In fact everything request to a Hail service (besides a lets-encrypt path) gets redirected to https.

> Or am I misunderstanding how this works?

Nope, you have it correct. $http_x_forwarded_proto should never be absent, and would be fine to use instead of $updated_scheme (but I'd prefer one of those two, rather than https, because otherwise we're not relying on our upstream infrastructure).

> * I'm having trouble understanding the difference between `Host` and `X-Forwarded-Host`, still. As I understand it, `Host` is the name of the server that the current request is trying to reach, and `X-Forwarded-Host` is the name of the server that the original request was trying to reach? Which is why `Host` is set to `$service.internal` and `X-Forwarded-Host` is `$http_host` in the internal.hail.is server? 

Yeah that's right. Host refers to the current server (or in the proxied case, what gateway set Host to). X-Forwarded-Host is set by gateway to be the $http_host at the time it proxies the request to router, which is going to be blog.hail.is.


> I don't quite follow your comment about our use of `Host` being wrong, in this case; I _think_ I understand what you're saying? but I'm not sure why all of our stuff is setting `Host` to `$updated_host` if that's the case, and I don't understand what's happening enough to know what happens if I change it to `proxy_set_header Host $http_host;` instead.

I just mean that we're setting Host to the value that X-Forwarded-Host has, when it seems more appropriate for it to be the domain:port of the internal request. So we're relying on X-Forwarded-Host to set that value, but we could, and I think it may be cleaner / more informative, to rely on X-Forwarded-Host all the way through to the internal server that's receiving the request - in this case Ghost's ExpressJS server - instead of rewriting the Host field to be that of the external request.

File: blog/deployment.yaml (hunk 8 above)
Comment (catoverdrive) [reply blog/deployment.yaml]: right---I guess my question is why we're setting the Host field to e.g. `blog.internal` here---what does that do?

Comment (akotlar) [reply blog/deployment.yaml]: Creates an internal hostname that the internal service could listen on (blog.internal).

Comment (akotlar) [reply blog/deployment.yaml]: Nginx uses the Host field to determine match server_name http://nginx.org/en/docs/http/request_processing.html

**Comment by akotlar** [comment]
Hm, the entry point that CI gets looks to be /

**Comment by akotlar** [comment]
Actually the endpoint does seem to be an issue.

https://internal.hail.is/pr-7381-default-sx9ail9zkm77/blog/ works great.

**Comment by akotlar** [comment]
@danking should internal_base_url in wait-for.py be https? Do we have anything that should route through http?

**Comment by akotlar** [comment]
Also, strangely https://internal.hail.is/pr-7381-default-sx9ail9zkm77/blog/ now goes to 403 Forbidden, and a maintenance error is generated in Ghost. Shortly after the PR was built, that link worked.

**Comment by danking** [comment]
PR namespaces are destroyed after the tests pass, so ghost should be gone.

**Comment by danking** [comment]
Or fail.

**Comment by akotlar** [comment]
Also note

> Or fail.

Ah, that makes sense.

**Comment by catoverdrive** [comment]
What do you mean by "a maintenance error is generated in Ghost"? I kind of assumed that the website wasn't working right now because the tests finished and `cleanup_deploy_blog` shut down the blog in the PR namespace.

The "endpoint" that I'm passing in to the `wait` command for the blog in build.yaml is `/`, which *should* do the right thing interally because the URL that's actually being constructed in `wait-for.py` becomes `http://{service}.{namespace}/{namespace}/{service}{endpoint}`, which gives you the right thing.

The test failure that I'm running into currently has to do with the fact that the wait command there queries the endpoint without going through either the router or the gateway, (since it's e.g. hitting http://blog.wang/wang/blog/ directly), so it's getting the 301 redirect to https because the `X-Forwarded-Proto` header isn't set. I'm not sure what the right fix is in this case.

**Comment by akotlar** [comment]
> What do you mean by "a maintenance error is generated in Ghost"? I kind of assumed that the website wasn't working right now because the tests finished and `cleanup_deploy_blog` shut down the blog in the PR namespace.
> 
> The "endpoint" that I'm passing in to the `wait` command for the blog in build.yaml is `/`, which _should_ do the right thing interally because the URL that's actually being constructed in `wait-for.py` becomes `http://{service}.{namespace}/{namespace}/{service}{endpoint}`, which gives you the right thing.
> 
> The test failure that I'm running into currently has to do with the fact that the wait command there queries the endpoint without going through either the router or the gateway, (since it's e.g. hitting http://blog.wang/wang/blog/ directly), so it's getting the 301 redirect to https because the `X-Forwarded-Proto` header isn't set. I'm not sure what the right fix is in this case.

[2019-10-30 20:03:15] [36mINFO[39m Ghost boot 5.169s
[2019-10-30 20:03:16] [31mERROR[39m "GET /pr-7381-default-sx9ail9zkm77/blog/" [31m503[39m 40ms
[31m
[31mSite is starting up, please wait a moment then retry.[39m

[1m[37mError ID:[39m[22m
    [90m4f8b8590-fb50-11e9-ab7c-9dd7e9eff310[39m

[90m----------------------------------------[39m

[90mMaintenanceError: Site is starting up, please wait a moment then retry.

**Comment by catoverdrive** [comment]
ah---that's just a log message from the readinessProbe sending requests before the server is fully up, I believe.

**Comment by catoverdrive** [comment]
(If you want to take a look at a live instance, I've deployed to my namespace and you can look at the batch logs here: https://ci.hail.is/batches/634 )

**Comment by akotlar** [comment]
> ah---that's just a log message from the readinessProbe sending requests before the server is fully up, I believe.

So partly it looks like that, especially with Ghost timing itself as >5s to actually listen on the port after the container starts up (or really after Ghost starts running which may be different).

The strange this is the timing. I thought the timestamps were supposed to correspond to the time the request was received/log message was generated (the log messages themselves can be printed out of order, but I think the timestamps should be accurate).

**Comment by danking** [comment]
Not sure what to do about wait for and ghost. rather silly of ghost to refuse connections on http when the base url is set to https

**Comment by akotlar** [comment]
> (If you want to take a look at a live instance, I've deployed to my namespace and you can look at the batch logs here: https://ci.hail.is/batches/634 )

looks great.

**Comment by akotlar** [comment]
> Not sure what to do about wait for and ghost. rather silly of ghost to refuse connections on http when the base url is set to https

I still don't quite understand why the wait command doesn't hit gateway. Isn't it just issuing an http request to `f'{base_url}/{endpoint)`? Why doesn't that hit gateway?

**Comment by danking** [comment]
Only external requests go to the gateway.

**Comment by akotlar** [comment]
> Only external requests go to the gateway.

Why doesn't this do that?

```
base_url = internal_base_url(namespace, name, port)
    async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=5.0)) as session:
        while True:
            try:
                async with session.get(f'{base_url}/healthcheck') as resp:
```

**Comment by catoverdrive** [comment]
http://blog.wang/wang/blog/, for instance, sends a request to the `blog` service in the `wang` namespace inside our k8s cluster, which is all handled by kubernetes. So the wait basically sends a GET request on the `/wang/blog/` endpoint directly to the `blog` service in that namespace, which never passes through either gateway or router.

**Comment by akotlar** [comment]
> http://blog.wang/wang/blog/, for instance, sends a request to the `blog` service in the `wang` namespace inside our k8s cluster, which is all handled by kubernetes. So the wait basically sends a GET request on the `/wang/blog/` endpoint directly to the `blog` service in that namespace, which never passes through either gateway or router.

Yeah, I got it, thanks. I think we may need to set the X-Forwarded-Proto proto in the wait_for.py http   command (session.get)

**Comment by catoverdrive** [comment]
I'm happy to make that change, but I'm not sure what the best way to do that would be. Hardcoding it seems less than ideal. Would it be reasonable to have another option to specify header flags to set in the `wait` step, similar to how we can specify custom endpoints?

**Comment by akotlar** [comment]
> Would it be reasonable to have another option to specify header flags to set in the `wait` step, similar to how we can specify custom endpoints?

Exactly, I think that's the right model. If headers are provided, set them.

File: blog/deployment.yaml (hunk 8 above)
Comment (catoverdrive) [reply blog/deployment.yaml]: oooh, that makes sense. Would it be reasonable to just omit setting the `Host` header? I see several other servers blocks don't do it.

Comment (akotlar) [reply blog/deployment.yaml]: Yeah I don't think we need it here, since Ghost won't use it when X-Forwarded-Host is set, as long as it trusts the upstream client, which it does by default.

**Comment by danking** [comment]
I think the YAML is most naturally a list of dicts, each dict having two fields key and value:
```
headers:
  - key: X-Forwarded-Proto
    value: https
```
CI will get `headers` as a list of dicts. Just pass those strings literally on the command line:
```
wait-for.py ... --header "key1: value1" --header "key2: value2" ...
```
If a user of CI wants to use characters that aren't valid in HTTP headers, that's their problem, not CIs or wait-for.

**Comment by akotlar** [comment]
> I think the YAML is most naturally a list of dicts, each dict having two fields key and value:
> 
> ```
> headers:
>   - key: X-Forwarded-Proto
>     value: https
> ```
> 
> CI will get `headers` as a list of dicts. Just pass those strings literally on the command line:
> 
> ```
> wait-for.py ... --header 'key1: value1' --header 'key2: value2' ...
> ```
> 
> If a user of CI wants to use characters that aren't valid in HTTP headers, that's their problem, not CIs or wait-for.

Why not wait-for.py ... --header "{key1: value1, key2: value2}" which will map precisely to the aiohttp headers field (which takes a dict)

**Comment by danking** [comment]
Just seems more complicated than using argparse to get repeated arguments. Actually looks like we can use nargs=2 to get: `--header 'key1' 'value1' --header 'key2' 'value2'`. Why introduce a JSON unparser and parser?

**Comment by akotlar** [comment]
> Just seems more complicated than using argparse to get repeated arguments. Actually looks like we can use nargs=2 to get: `--header 'key1' 'value1' --header 'key2' 'value2'`. Why introduce a JSON unparser and parser?

Up to you. JSON parser/unparser is stdlib, and it looks like the command line tool is little used (meaning, executed typically as part of a larger script). So I'm optimizing for the typical use case (it's easier to specify the YAML property using the dict method, and easier to write/maintain the parser, since no need to go from array to dict). Neat about args 2.

**Comment by akotlar** [comment]
needs bump
